docs/articles/improve-raptor-with-rag.md

```diff
@@ -0,0 +1,561 @@
+
+**Setup and Imports**
+
+```python
+pip install lancedb scikit-learn openai torch sentence_transformers tiktoken umap-learn PyPDF2
+```
+
+```python
+import os
+import uuid
+import tiktoken
+import re
+import numpy as np
+import pandas as pd
+import transformers
+import torch
+import umap.umap_ as umap
+import matplotlib.pyplot as plt
+from openai import OpenAI
+from typing import List, Tuple, Optional, Dict
+from sklearn.mixture import GaussianMixture
+from sentence_transformers import SentenceTransformer
+
+openai_api_key = "sk-XXXXXXXXXXXXXXX"
+client = OpenAI(api_key=openai_api_key)
+```
+
+### Creating the Chunks
+
+Setting up RAPTOR is pretty straightforward and builds on what we’ve already covered. The first step is to break down our textual documents into smaller chunks. Once we have those, we can convert them into dense vector embeddings.
+
+```python
+import os
+import PyPDF2
+
+# Function to extract text from a PDF file
+def extract_pdf_text(file_path):
+    with open(file_path, 'rb') as file:
+        reader = PyPDF2.PdfReader(file)
+        text = ""
+        for page_num in range(len(reader.pages)):
+            page = reader.pages[page_num]
+            text += page.extract_text()
+    return text
+
+# Function to split text into chunks with overlap
+def split_text(text, chunk_size=1000, chunk_overlap=50):
+    chunks = []
+    start = 0
+    while start < len(text):
+        end = min(start + chunk_size, len(text))
+        chunks.append(text[start:end])
+        start += chunk_size - chunk_overlap
+    return chunks
+
+# Function to process all PDFs in a directory
+def process_directory(directory_path, chunk_size=1000, chunk_overlap=50):
+    all_chunks = []
+    # Iterate over all PDF files in the directory
+    for filename in os.listdir(directory_path):
+        if filename.endswith(".pdf"):
+            file_path = os.path.join(directory_path, filename)
+            print(f"Processing file: {file_path}")
+            
+            # Step 1: Extract text from the PDF
+            pdf_text = extract_pdf_text(file_path)
+            
+            # Step 2: Split the extracted text into chunks
+            chunks = split_text(pdf_text, chunk_size=chunk_size, chunk_overlap=chunk_overlap)
+            
+            # Append chunks from this file to the overall list
+            all_chunks.extend(chunks)
+    
+    return all_chunks
+
+directory_path = os.path.join(os.getcwd(), "data")  # Specify your directory path
+chunk_size = 1000
+chunk_overlap = 50
+
+# Process all PDF files in the directory and get the chunks
+chunks = process_directory(directory_path, chunk_size=chunk_size, chunk_overlap=chunk_overlap)
+
+# Optional: Print the number of chunks and preview some of the chunks
+print(f"Total number of chunks: {len(chunks)}")
+for i, chunk in enumerate(chunks[:5]):  # Show first 5 chunks as a preview
+    print(f"Chunk {i+1}:\n{chunk}\n")
+```
```

Comment:
can you provide the number of chunks and a preview here then?

Reply:
This looks good! I mean there is nothing that needs to be changed, Let me know if this sounds ok!

Reply:
I think we don't need to show the chunks here, I just put it up during the testing.

Thanks for pointing it out @robertdhayanturner.

Reply:
in our notebook articles, typically we include output. Your article has only two outputs at the very bottom. If there's no reason to not include it, and since it's straightforward, I would include it.
I just tried running your collab. Ran into error.

<img width="829" alt="Screenshot 2024-10-21 at 16 51 19" src="https://github.com/user-attachments/assets/568c27ca-f2e1-43eb-a6bf-9ba413931b69">

Reply:
The breaking down the data into chunks step results in 0 chunks.

Reply:
![image](https://github.com/user-attachments/assets/e92b9078-4129-4a09-bcc5-bafd34ac4573)

Make sure you have a PDF in the data folder!

Reply:
I am running the same script and it's working fine and good!

![image](https://github.com/user-attachments/assets/7c5a9e00-d570-4189-b5fd-e7193a342670)

Reply:
Added a note to make sure people don't forget to add the PDFs before running the script!

Reply:
Where do I find the pdf/s so I can add them?
And... at this point, does your colab user know what the file is?

Reply:
This is simply RAG (With RAPTOR which is in turn an advanced strategy to improve RAG). You can just upload any relevant PDFs that you want to perform Retrieval-Augmented Generation (RAG) on, and then run the provided Colab notebook as-it is.

Reply:
Specifically, in the conclusion you mention:
"RAPTOR RAG performs better than vanilla RAG. RAPTOR retrieves specific details about NTT's financial growth for the specified quarter, and  connects this growth to the broader acquisition strategy, pulling relevant context from multiple sources."
in fact you have one source.
I will make the corrections throughout.

Reply:
> Added a note to make sure people don't forget to add the PDFs before running the script!

this entire snippet is missing from the article:

<img width="1006" alt="Screenshot 2024-10-22 at 14 28 51" src="https://github.com/user-attachments/assets/0ca555a1-abbd-4ea5-815c-a50b9885df4d">

Reply:
Uhh, can we have a quick call on slack? Let me know a suitable time and we will sort this out, I am still confused how you are getting this error because I sent this colab to more than 5 individuals and everyone was able to run it smoothly!

Slacked you!

Reply:
Well for the "Multiple sources" I wanted to say "Multiple Chunks (old chunks + chunks which are formed with the RAPTOR) 

Thanks for the clarification Robert! Means a ton!

Reply:
thanks again for the huddle - 
resolved issues in my colab:
1) encrypted file upload to '/content/data'; Vipul sent unencrypted file replacement; error solved
2) rate limit error (no funds in chatgpt acct; solved

closing this issue